src/awkward/_v2/contents/numpyarray.py

```diff
@@ -581,13 +581,20 @@ def _as_unique_strings(self, offsets):
         return out2, nextoffsets[: outlength[0]]
 
     def numbers_to_type(self, name):
-        dtype = primitive_to_dtype(name)
-        return NumpyArray(
-            self._nplike.asarray(self._data, dtype=dtype),
-            self._identifier,
-            self._parameters,
-            self._nplike,
-        )
+        if (
+            self.parameter("__array__") == "string"
+            or self.parameter("__array__") == "bytestring"
+            or self.parameter("__array__") == "char"
+        ):
+            return self
```

Comment:
```suggestion
        if (
            self.parameter("__array__") == "string"
            or self.parameter("__array__") == "bytestring"
            or self.parameter("__array__") == "char"
            or self.parameter("__array__") == "byte"
        ):
            return self
```

These kinds of checks may be proliferating, so they should probably become a utility function in `ak._v2._util`. That would best be handled in another PR, though.